### ISSUE TITLE
support inert attribute

### ISSUE THREAD

**Issue by idoros** [focus-trap/tabbable#292]
Elements that are marked with [inert](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/inert) or are nested in another `inert` element shouldn't be tabbable/focusable. Since this is just supported behind browsers flags at the moment, maybe tabbable can add support with a default false flag for now.

**Comment by stefcameron** [comment]
Still behind a flag a year later.

**Comment by stefcameron** [comment]
I'd prefer to wait until it has at least some real traction. Closing for now.

**Comment by KittyGiraudel** [comment]
[Inert is here](https://caniuse.com/?search=inert). :)

**Comment by stefcameron** [comment]
The [inert attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert) does have reasonable support now, thanks for pointing that out (though curiously still nothing from Firefox).

I was initially thinking that supporting this tag in terms of tabbable should "only" require updating the [candidateSelector](https://github.com/focus-trap/tabbable/blob/master/src/index.js#L1-L14) to be:

```javascript
const candidateSelectors = [
  '*:not([inert]) input',
  '*:not([inert]) select',
  '*:not([inert]) textarea',
  '*:not([inert]) a[href]',
  '*:not([inert]) button',
  '*:not([inert]) [tabindex]:not(slot)',
  '*:not([inert]) audio[controls]',
  '*:not([inert]) video[controls]',
  '*:not([inert]) [contenteditable]:not([contenteditable="false"])',
  '*:not([inert]) details>summary:first-of-type',
  '*:not([inert]) details',
];
```

Unfortunately, it doesn't work.

If I have a DOM like:
```html
<div inert>
  <input />
</div>
```

The input will still match, which I don't understand given `*:not([inert])` should match any element that does _not_ have an `inert` attribute (right?) and then a space and then `input` should finally match any input that is a _descendant_ of any parent without an `inert` attribute. But I'm somehow mistaken here. I think we've hit a similar type of limitation on use of selectors like this for something in the past, but I don't recall exactly what case that was.

If I change the input selector to `*:not([inert]) > input` then the input doesn't match, but this obviously breaks as soon as the input is deeper down (say wrapped in another div).

So why doesn't this work for descendants at any level below the parent?

Of course, the opposite seems to work fine, `*[inert] input` will match it no matter how deep it is.

What I'm worried about is performance. At first glance, it seems tabbable would have to search up the hierarchy looking for an inert parent for every candidate it finds, which could slow things down considerably, especially since we'd be doing this all the time.

Granted we do this type of exhaustive check depending on options like `getShadowRoot` (which invokes `getCandidatesIteratively()` instead of just `getCandidates()`) but that's an opt-in if you're using web components/shadow DOM, which the case most of the time with this library.

I guess there might be a solution where we have two sets of candidate selectors: One which is the existing set that finds everything, and another that finds all things inert:

```javascript
const inertCandidateSelectors = [
  '*[inert] input',
  '*[inert] select',
  '*[inert] textarea',
  '*[inert] a[href]',
  '*[inert] button',
  '*[inert] [tabindex]:not(slot)',
  '*[inert] audio[controls]',
  '*[inert] video[controls]',
  '*[inert] [contenteditable]:not([contenteditable="false"])',
  '*[inert] details>summary:first-of-type',
  '*[inert] details',
];
```

And then for every candidate we find, we check (much faster) if it's in the inert list, and if so, exclude it.

Thoughts?

@idoros WDYT?

If there's a reasonable viable performant-looking solution, I'll re-open this and get it done.

**Comment by KittyGiraudel** [comment]
The reason `:not([inert]) input` doesn’t work the way you expect it to work is the same reason `:not(A) B` doesn’t. `:not(A)` is the same as `*:not(A)` which matches every node on the page which is not `A`. It’s likely one of them higher upper does match.

Take the following very reduced HTML document:

```html
<html>
  <style>:not(body) div { background-color: red }</style>
  <body>
    <div></div>
  <body>
</html>
```

At first glance, you’d expect the `div` **not** to have a red background. After all, it **is** inside the body. The thing is, `:not(body) div` is — in that case — semantically equivalent to: `html:not(body) div, body:not(body) div, div:not(body) div`. And of course, the 1st and the 3rd do match.

**Comment by idoros** [comment]
@KittyGiraudel is correct about your selector problem, any ancestor that doesn't have `inert` will match, ignoring the one that does.

> When walking the tree to find tabbable elements, we discard any sub-tree that’s inert (or hidden or disabled).

The process that picks tabbable elements is implemented twice, one for the default (maybe it should be referred as "legacy"? 😉) that doesn't support shadow-dom and doesn't rely on JavaScript tree traversal, and the new once which traverse the DOM and could simply stop when reaching `inert` (although it would still require an initial check to see that the starting element doesn't descend from an inerted element).

The "legacy" process would require another check that any candidate element isn't a nested under `inert`.

In addition the same test is required for the `isFocusable()` and `isTabbable()` APIs.

**Comment by stefcameron** [comment]
@KittyGiraudel Thank you for the CSS explainer, very helpful.

@idoros Thank you for the pointers.

Seems reasonable so I will reopen this and will try to get it done in the next few weeks when I have cycles again.

**Comment by idoros** [comment]
Here is a way to solve this with just a selector for the candidates:

```css
input:not([inert] *)
```

**Comment by idoros** [comment]
Will work

**Comment by stefcameron** [comment]
@idoros @KittyGiraudel #924 adds support for inert. Please have a look and LMK if you think I missed something. In the end, I had to use a combination of selector and manual checks because I found the `:not([inert] *)` selector, while useful, still doesn't catch all things depending on the entry point.

Also, as `inert` is not supported on FF, I've had to disable the repo's requirement for FF tests to pass when merging PRs. Hopefully FF gets on board soon... For now, I'll have to manually make sure that the only failures on FF are the inert tests...

**Comment by stefcameron** [comment]
Published in v6.1.0

**Comment by stefcameron** [comment]
Incidentally, JSDom does not support `:not([inert] *)` (but it does support `:not([inert])`, thankfully), so v6.1.0 broke the Internet... It also doesn't support the `HTMLElement.inert` API, so had to fix that too. #926
